assets/js/react/GameStateEditor/components/EntityForm/EntityForm.jsx

```diff
@@ -46,132 +51,249 @@ class EntityForm extends Component {
     this.setState({
-      entity: {
-        ...currentState.entity,
-        behaviors: map(prop("value"), behaviors),
-      },
+      behaviors: omit([ev.target.name], this.state.behaviors),
     })
   }
 
-  handleChangePropsDropdown = fieldName => {
-    return selectedValue => {
-      const currentState = this.state
-      this.setState({
-        entity: {
-          ...currentState.entity,
-          props: {
-            ...currentState.entity.props,
-            [fieldName]: selectedValue.value,
-          },
-        },
+  handleAddNewBehavior = selectedOption => {
+    const currentState = this.state
+    this.setState(
+      mergeDeepRight(currentState, {
+        behaviors: { [selectedOption.value]: {} },
       })
-    }
+    )
   }
 
+  handleChangeLocation = selectedOption =>
+    this.setFieldState("props.location", selectedOption.value)
+
   handleChangeName = ev => {
     const currentState = this.state
     const newName = ev.target.value || ""
-    this.setState({
-      entity: {
-        ...currentState.entity,
+    console.log("change name", ev, currentState, newName)
+    this.setState(
+      mergeDeepRight(currentState, {
         id: newName
           .trim()
           .replace(/[^\w\d]+/gi, "-")
           .toLowerCase(),
         name: newName,
-      },
-    })
+      })
+    )
   }
 
-  handleChange = ev => {
-    const currentState = this.state
-    this.setState({
-      entity: {
-        ...currentState.entity,
-        [ev.target.name]: ev.target.value,
-      },
-    })
+  getFieldState = fieldName => {
+    return get(this.state, fieldName)
   }
 
-  handleChangeProps = ev => {
+  setFieldState = (fieldName, value) => {
     const currentState = this.state
-    this.setState({
-      entity: {
-        ...currentState.entity,
-        props: {
-          ...currentState.entity.props,
-          [ev.target.name]: ev.target.value,
-        },
-      },
-    })
+    const nameBits = fieldName.split(".")
+    const update = assocPath(nameBits, value)({})
+    this.setState(mergeDeepRight(currentState, update))
   }
 
+  handleChange = ev => this.setFieldState(ev.target.name, ev.target.value)
+
+  handleChangeDestination = fieldName => selectedOption =>
+    this.setFieldState(fieldName, selectedOption.value)
+
   handleAdd = ev => {
     ev.preventDefault()
-    this.props.addEntity(this.state.entity)
+    this.props.addEntity(this.state)
   }
 
   handleClear = _ev => {
-    this.setState({ entity: null })
+    this.setState(emptyEntity)
   }
 
   valid = () => {
-    const requiredPaths = [["id"], ["props", "description"]]
-    return all(pth => path(pth, this.state.entity), requiredPaths)
+    const requiredPaths = ["id", "name"]
+    return (
+      all(requiredPaths, pth => !isEmpty(get(this.state, pth))) &&
+      !isEmpty(Object.keys(this.state.behaviors))
+    )
+  }
+
+  determineFieldType = (fieldName, value) => {
+    if (fieldName.match(/\.destinationId$/)) {
```

Comment:
Here is where we could (if necessary) build custom form fields for custom behaviors.  

Until that's needed, we have "destination" or text field.  And without much effort, this could do booleans or whatever as needed.

As mentioned, turntable which wants an array of rewards is not yet handled, but this is where we could do that.